core/App/ffi_operation.cpp

```diff
@@ -305,12 +308,96 @@ extern "C"
 
         export_json_from_struct(retJsonObj, master_key, "cmk");
     out:
-        SAFE_FREE(key_metadata); 
+        SAFE_FREE(key_metadata);
         SAFE_FREE(master_key);
         retJsonObj.toChar(respJson);
         return ret;
     }
 
+    /**
+     * @brief Get public key from sm2 keypair
+     *
+     * @param payload : Pass in the key parameter in the form of JSON string
+                {
+                    cmk : a base64 string,
+                }
+     *
+     * @return char*
+     * [string] json string
+        {
+            code: int,
+            message: string,
+            result: {
+                ciphertext : a base64 string
+            }
+        }
+     */
+    uint32_t ffi_getPublicKey(JsonObj payloadJson, char *respJson)
+    {
+        RetJsonObj retJsonObj;
+        ehsm_status_t ret = EH_OK;
+
+        ehsm_keyblob_t *cmk = NULL;
+        ehsm_data_t *pubkey = NULL;
+
+        JSON2STRUCT(payloadJson, cmk);
+
+        if (cmk == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Invalid Parameter.");
+            goto out;
+        }
+
+        pubkey = (ehsm_data_t *)malloc(APPEND_SIZE_TO_DATA_T(0));
+        if (pubkey == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+        pubkey->datalen = 0;
+
+        ret = GetPublicKey(cmk, pubkey);
+        if (ret != EH_OK)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+
+        if (pubkey->datalen == 0 || pubkey->datalen > UINT16_MAX)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+
+        pubkey = (ehsm_data_t *)realloc(pubkey, APPEND_SIZE_TO_DATA_T(pubkey->datalen));
+        if (pubkey == NULL)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+
+        ret = GetPublicKey(cmk, pubkey);
+        if (ret != EH_OK)
+        {
+            retJsonObj.setCode(retJsonObj.CODE_FAILED);
+            retJsonObj.setMessage("Server exception.");
+            goto out;
+        }
+
+        STRUCT2JSON(retJsonObj, pubkey);
```

Comment:
It's encoded by b64. try it:
retJsonObj.addData_string("PublicKey", pubkey->data);